assignments/introductory-assignments.md

```diff
@@ -488,14 +582,767 @@ A digital circuit computer with 2GB of memory can distinguish 2^{17 179 869 184}
+
+We take:
+
+- the opcode 19 (`0010011`),
+- register 17 (`10001`),
+- register 0 (`00000`),
+- and value 214 (`000011010110`) 
+ 
+and put them together in binary as follows:
+
+`000011010110 00000 000 10001 0010011`
+
+If you merge that into a 32-bit word, you get `0x0D600893`.
+
+The RISC-V ISA specifies that:
+
+- the twelve MSBs encode the third (!) operand value 214, in 12-bit two's complement in fact, so it could even be a negative value.
+- The next five bits encode the second (!) operand register 0, followed by three bits set to 0.
+- The next five bits encode the first (!) operand register 17 followed by,
+- the remaining seven LSBs that encode the opcode 19.
+
+The third operand is treated by `addi` as an integer value rather than, say, a number identifying a register, is called *immediate* addressing hence the `i` in `addi`. Immediate addressing is one of various so-called *addressing modes*.
+
+[Addressing Mode](https://en.wikipedia.org/wiki/Addressing_mode) An aspect of the instruction set architecture in most central processing unit (CPU) designs. The various addressing modes defined in a given instruction set architecture to define how machine language instructions in that architecture identify each instruction's operand(s).
 
-Since the states (i.e. 0 and 1) are represented by different voltage levels on the electrical level, it is better to have as few states as possible. The reasons for this are based on physics: By things like heat or other electrical disturbances the voltage levels can be changed easily. So it is easier to compensate for these fluctuations with two different states than it would be with, say, ten states - which would correspond to the decimal system normally used by humans. In order to be able to distinguish between all states, one would have to increase the generally possible voltage, which would lead to an increased power consumption. Additionally the logic, which is needed on hardware level to calculate with multiple states, would be much more complicated than it is the case with only two states.
+But why does the ISA provision five bits for the first and second operand? Because five bits allow us to address exactly the 2^{5}=32 different registers of the machine. The seven bits for the opcode allow us to distinguish up to 2^{6}=64 different opcodes, but we do not need that many. Now, think about the range of values that can be encoded in two's complement in the twelve bits for the third operand! This is the range of values that we can get into a register such as `$a7` with a single `addi` instruction. In other words, we can use that instruction to initialize registers. 
```

Comment:
```suggestion
But why does the ISA provision five bits for the first and second operand? Because five bits allow us to address exactly the 2^{5}=32 different machine registers. The opcode's seven bits allow us to distinguish up to 2^{6}=64 different opcodes, but we do not need that many. Now, think about the range of values that can be encoded in two's complement in the twelve bits for the third operand! This is the range of values that we can get into a register, such as `$a7` with a single `addi` instruction. In other words, we can use that instruction to initialize registers. 
```